### PR TITLE
[APB-1125][JR,ZF] /registrations GET endpoint now extracts registrations

### PR DESCRIPTION
**Pull request by johnrevill**

### CODE REVIEW

File: app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala [hunk 1]
```diff
@@ -37,23 +38,43 @@ class AgentEpayeRegistrationRepository @Inject()(mongo: ReactiveMongoComponent)
...
-      _ <- insert(RegistrationDetails(nextAgentRef, request)) recover {
+      _ <- insert(RegistrationDetails(nextAgentRef, request, createdDate)) recover {
         case error: DatabaseException if error.code.contains(mongoCodeDuplicateKey) => create(request)
       }
     } yield nextAgentRef
   }
+
+  def findRegistrations(dateTimeFrom: DateTime, dateTimeTo: DateTime)
+                       (implicit ec: ExecutionContext): Future[List[RegistrationDetails]] = {
+    if(dateTimeTo.isBefore(dateTimeFrom)) throw new IllegalArgumentException("to date is before from date")
```
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala]: You can use the ```require()``` statement here.

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala [hunk 2]
```diff
@@ -73,8 +77,23 @@ object AgentEpayeRegistrationValidator {
       request.address.addressLine4.map(x => validCharsWithLimit(x, "address line 4", 35))
     ).flatten
 
+    validate(validators ++ optionalFieldValidators)
+  }
+
+  def validateDateRange(dateFrom: String, dateTo: String): Validated[Failure, Unit] = {
+    def dateInPast(param: String, paramName: String) =
+      isISODate(param)(paramName).andThen(_ => isInPast(parseISODate(param))(paramName))
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: Do we have to parse the same date multiple times? Maybe we could use PathBinders to have endpoint parameters of `Date` type?

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala [hunk 3]
```diff
@@ -113,4 +132,32 @@ object AgentEpayeRegistrationValidator {
...
+  private[validators] def isInPast(date: LocalDate)(paramName: String) = {
+    val today = LocalDate.now(DateTimeZone.UTC)
+    if (today.isAfter(date))
+      Valid(())
+    else
+      Invalid(Failure("INVALID_DATE_RANGE", s"'$paramName' date must be in the past"))
+  }
+
+  private[validators] def isValidDateRange(dateFrom: LocalDate, dateTo: LocalDate) = {
+    if (dateFrom.isAfter(dateTo))
```
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: `isAfter` may throw an exception in some cases. Shouldn't we consider using Try here too?

File: app/uk/gov/hmrc/agentepayeregistration/models/RegistrationExtraction.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    import details._
+    import details.registration._
+    import details.registration.address._
+    RegistrationExtraction(agentReference,
+      agentName,
+      contactName,
+      telephoneNumber,
+      faxNumber,
+      emailAddress,
+      addressLine1,
```
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/models/RegistrationExtraction.scala]: Shouldn't we need address in a separate object? The post endpoint uses such format as input json.

Comment (arturopala) [reply app/uk/gov/hmrc/agentepayeregistration/models/RegistrationExtraction.scala]: I think we better keep it flat here to match expected output json.

File: app/uk/gov/hmrc/agentepayeregistration/controllers/AgentEpayeRegistrationController.scala [hunk 5]
```diff
@@ -44,16 +44,17 @@ class AgentEpayeRegistrationController @Inject()(@Named("extract.auth.stride.enr
...
-        case ex: AuthorisationException =>
-          Future.successful(Forbidden(ex.getMessage))
+      service.extract(dateFrom, dateTo).map {
+        case Right(registrations) => Ok(Json.obj("registrations" -> Json.toJson(registrations)))
+        case Left(failure) => BadRequest(Json.toJson(failure))
       }
-
+    }.recoverWith {
+      case ex: NoActiveSession =>
+        Future.successful(Unauthorized(ex.getMessage))
```
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/controllers/AgentEpayeRegistrationController.scala]: Not sure we need to expose the error message here, should we only need to log the message?

File: app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala [hunk 6]
```diff
@@ -37,23 +38,43 @@ class AgentEpayeRegistrationRepository @Inject()(mongo: ReactiveMongoComponent)
...
+      "createdDateTime" -> obj(
+        "$gte" -> obj("$date" -> dateTimeFrom.getMillis),
+        "$lte" -> obj("$date" -> dateTimeTo.getMillis)
+      )
+    )
+
+    collection.find(queryFilter)
+      .sort(obj("createdDateTime" -> 1))
+      .cursor[RegistrationDetails]()
+      .collect[List](Integer.MAX_VALUE)
```
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala]: You could try
```scala
 find("createdDateTime" -> obj(
      "$gte" -> obj("$date" -> dateTimeFrom.getMillis),
      "$lte" -> obj("$date" -> dateTimeTo.getMillis)
    )).map(_.sortBy(_.createdDateTime)(Ordering.fromLessThan(_ isBefore _)))
```

The find method is from `uk.gov.hmrc.mongo.ReactiveRepository` class

Comment (ZDevelop94) [reply app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala]: that method is expecting a JsValueWrapper)* which is quite lengthy to change.

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala (hunk 2 above)
Comment (johnrevill) [reply app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: The `dateFrom` and `dateTo` parameters now use a `QueryStringBinder` for `LocalDates`. This cleans up the parsing/validation a little bit, at the slight expense of a different JSON error response format when date parameters fail to parse (as it's generated by Play)

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala [hunk 7]
```diff
@@ -113,4 +128,26 @@ object AgentEpayeRegistrationValidator {
...
+  private[validators] def isInPast(date: LocalDate)(paramName: String) = {
+    val today = LocalDate.now(DateTimeZone.UTC)
+    if (today.isAfter(date))
+      Valid(())
+    else
+      Invalid(Failure("INVALID_DATE_RANGE", s"'$paramName' date must be in the past"))
+  }
+
+  private[validators] def isValidDateRange(from: LocalDate, to: LocalDate) = {
+    Try(from.isAfter(to)) match {
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: I think we could simplify this again assuming date is never null, no exception is thrown? wdyt?

File: app/uk/gov/hmrc/agentepayeregistration/binders/Bindings.scala [hunk 8]
```diff
@@ -0,0 +1,38 @@
+/*
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentepayeregistration/binders/Bindings.scala]: we have some previous examples of putting binders into controller/UrlBinders, maybe better stick with it: https://github.com/hmrc/agent-mapping/blob/0ba136e00eebb556baea23d94c5a82a434744df5/app/uk/gov/hmrc/agentmapping/controller/UrlBinders.scala, https://github.com/hmrc/agent-subscription/blob/f869fd7a1d540dc2efd1f947cdd310c2e7a93fa3/app/uk/gov/hmrc/agentsubscription/controllers/UrlBinders.scala

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala (hunk 7 above)
Comment (johnrevill) [reply app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: Sure, I'll change it back. It was previously making that assumption, but I changed it after another PR comment mentioned `isAfter` could throw an exception. The only cases where it'll throw an exception are when the parameter is null though.

File: app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala [hunk 9]
```diff
@@ -37,23 +37,37 @@ class AgentEpayeRegistrationRepository @Inject()(mongo: ReactiveMongoComponent)
...
       }
-      _ <- insert(RegistrationDetails(nextAgentRef, request)) recover {
+      _ <- insert(RegistrationDetails(nextAgentRef, request, createdDate)) recover {
         case error: DatabaseException if error.code.contains(mongoCodeDuplicateKey) => create(request)
       }
     } yield nextAgentRef
   }
+
+  def findRegistrations(dateTimeFrom: DateTime, dateTimeTo: DateTime)
+                       (implicit ec: ExecutionContext): Future[List[RegistrationDetails]] = {
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala]: Would be better to stream response here, wdyt?

File: app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala [hunk 10]
```diff
@@ -37,23 +37,37 @@ class AgentEpayeRegistrationRepository @Inject()(mongo: ReactiveMongoComponent)
...
+    require(!dateTimeTo.isBefore(dateTimeFrom), "to date is before from date")
+
+    val queryFilter: (String, JsValueWrapper) = "createdDateTime" -> obj(
+        "$gte" -> obj("$date" -> dateTimeFrom.getMillis),
+        "$lte" -> obj("$date" -> dateTimeTo.getMillis))
+
+    find(queryFilter).map(_.sortBy(_.createdDateTime)(Ordering.fromLessThan(_ isBefore _)))
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala]: can we sort on mongodb side?

**Pull request merged**